src/main/kotlin/org/neo4j/graphql/GraphQLExtensions.kt

```diff
@@ -46,13 +46,18 @@ fun GraphQLFieldsContainer.relationshipFor(name: String): RelationshipInfo? {
             ?: throw IllegalArgumentException("$name is not defined on ${this.name}")
     val fieldObjectType = field.type.inner() as? GraphQLFieldsContainer ?: return null
 
-    // TODO direction is depending on source/target type
-
-    val (relDirective, isRelFromType) = (fieldObjectType as? GraphQLDirectiveContainer)
-        ?.getDirective(DirectiveConstants.RELATION)?.let { it to true }
-            ?: field.getDirective(DirectiveConstants.RELATION)?.let { it to false }
-            ?: throw IllegalStateException("Field $field needs an @relation directive")
+    val (relDirective, isRelFromType) = if (isRelationType()){
+        (this as? GraphQLDirectiveContainer)
+            ?.getDirective(DirectiveConstants.RELATION)?.let { it to false }
+                ?: throw IllegalStateException("Field $field needs an @relation directive")
+    } else {
+        (fieldObjectType as? GraphQLDirectiveContainer)
+            ?.getDirective(DirectiveConstants.RELATION)?.let { it to true }
+                ?: field.getDirective(DirectiveConstants.RELATION)?.let { it to false }
```

Comment:
shouldn't in the case that the object type is a relationship-type it fail as well if there is no @relationship directive on the type? it seems to fail only when it's missing on the field, but AFAIR the one on the field is optional then and only needed when it is a same-type relationship, not sure.

Reply:
This is the old logic, I just added the branch for handling relation types. I'll adjust the message of the Exception for the new branch.